src/genotyper.cc

```diff
@@ -41,6 +41,286 @@ static bool is_pseudo_ref_record(const bcf_hdr_t* hdr, bcf1_t* record) {
+                return Status::Invalid("genotyper: getting format field errored with type mismatch", errmsg.str());
+            }
+            // don't raise error if get failed due to tag missing in record or
+            // in vcf header; continue to look with other possible field names
+
+            if (rv >= 0) {
+                found = true;
+                if (rv != record->n_sample * n_val_per_sample) {
+                // For this field, we expect n_val_per_sample values per sample
+                    if (v) free(v);
+                    ostringstream errmsg;
+                    errmsg << dataset << " " << range(record).str() << "(" << field_name << ")";
+                    return Status::Invalid("genotyper: unexpected result when fetching record FORMAT field", errmsg.str());
+                } // close rv != record->n_sample * count
+
+                for (int i=0; i<record->n_sample; i++) {
+                    for (int j=0; j<n_val_per_sample; j++) {
+
+                        int in_ind = i * n_val_per_sample + j;
+                        int out_ind = get_out_ind_of_value(i, j, sample_mapping, allele_mapping);
+
+                        if (out_ind < 0) {
+                           continue;
+                        }
+
+                        assert(out_ind < format_v.size());
+                        assert(in_ind < vsz);
+                        format_v[out_ind].push_back(v[in_ind]);
+                    } // close for j loop
+                } // close for i loop
+            } // close rv >= 0
+            free(v);
+        }
+
+        if (!found) {
+
+            // Special case handling for AD field to convert ref
+            // DP/MIN_DP to repopulate AD field
+            if (field_info.name == "AD" && !handled_AD_field) {
+
+                // Prevent runaway recursion
+                handled_AD_field = true;
+
+                // Call add_record_data again, searching for DP, MIN_DP, override n_val_per_sample to 1
+                Status s = add_record_data(dataset, dataset_header, record, sample_mapping, allele_mapping, {"DP", "MIN_DP"}, 1);
+
+                // Reset flag for next dataset
+                handled_AD_field = false;
+
+                return s;
+            }
+            // TODO: The specified field could not be found from the input record, this is "OK"
+            // because we might be able to find some other record that has the requisite value
+            // Populating by default value is handled in the update_record_format step
+            return Status::OK();
+        }
+        return Status::OK();
+    }
+
+    Status update_record_format(const bcf_hdr_t* hdr, bcf1_t* record) {
+        Status s;
+        vector<T> ans;
+        s = combine_format_data(ans);
+
+        if (!s.ok()) {
+            // TODO: Combine function failed; currently ignore error, do not
+            // update this field, and move on. More intelligent error handling?
+            return Status::OK();
```

Comment:
yes, the upstream logic in combine_format_data may not be the best... I changed the upstream logic in combine_format_data to not fail when the format field is unrepresentable in the given record (e.g. when there's a missing value), but only raise an error if there's actually an issue... so we can return s if s is not ok here..